src/nomad_parser_magres/parsers/parser.py

```diff
@@ -0,0 +1,740 @@
+        if input_structure:
+            workflow.m_add_sub_section(
+                NMRMagRes.inputs, Link(name='Input structure', section=input_structure)
+            )
+        if nmr_magres_calculation:
+            workflow.m_add_sub_section(
+                NMRMagRes.outputs,
+                Link(name='Output NMR calculation', section=nmr_magres_calculation),
+            )
+
+        # NMR (first principles) task
+        # ! Fix this once CASTEP and QuantumESPRESSO use the new `nomad-simulations` schema under 'data'
+        program_name = nmr_first_principles_archive.run[-1].program.name
+        if nmr_first_principles_archive.workflow2:
+            task = TaskReference(task=nmr_first_principles_archive.workflow2)
+            task.name = f'NMR FirstPrinciples {program_name}'
+            if input_structure:
+                task.inputs = [Link(name='Input structure', section=input_structure)]
+            if nmr_magres_calculation:
+                task.outputs = [
+                    Link(
+                        name='Output NMR calculation',
+                        section=nmr_magres_calculation,
+                    )
+                ]
+            workflow.m_add_sub_section(NMRMagRes.tasks, task)
+
+        self.archive.workflow2 = workflow
+
+    def parse(
+        self, filepath: str, archive: 'EntryArchive', logger: 'BoundLogger'
+    ) -> None:
+        self.mainfile = filepath
+        self.maindir = os.path.dirname(self.mainfile)
+        self.basename = os.path.basename(self.mainfile)
+        self.archive = archive
+
+        self.init_parser(logger=logger)
+        self._check_units_magres(logger=logger)
+
+        # Adding Simulation to data
+        simulation = Simulation()
+        calculation_params = self.magres_file_parser.get('calculation', {})
+        if calculation_params.get('code', '') != 'CASTEP':
+            logger.error(
+                'Only CASTEP-based NMR simulations are supported by the magres parser.'
+            )
+            return
+        simulation.program = Program(
+            name=calculation_params.get('code', ''),
+            version=calculation_params.get('code_version', ''),
+        )
+        archive.data = simulation
+
+        # `ModelSystem` parsing
+        model_system = self.parse_model_system(logger=logger)
+        if model_system is not None:
+            simulation.model_system.append(model_system)
+
+        # `ModelMethod` parsing
+        model_method = self.parse_model_method(calculation_params=calculation_params)
+        simulation.model_method.append(model_method)
+
+        # `Outputs` parsing
+        outputs = self.parse_outputs(simulation=simulation, logger=logger)
+        if outputs is not None:
+            simulation.outputs.append(outputs)
+
+        # Try to resolve the `entry_id` and `mainfile` of other entries in the upload to connect the magres entry with the CASTEP or QuantumESPRESSO entry
```

Comment:
In the metadata section of file upload, there is a files section that lists all files included in the upload. Is this the only intended result? There is mention of linking the NMR workflow directly in the magres entry, could you please clarify how it is done? As per my previous comments, we see two separate file entries and workflows for the .castep and .magres files.